File: examples/native/macos/OneAIBusClient.swift
// OneAIBusClient.swift
// macOS native frontend for `oneai serve` (the engine-bus sidecar).
//
// A Directive writer + Yield reader over a Unix domain socket. This replaces
// the extern-C FFI binding (crates/oneai-platform-desktop/macos.rs +
// bridge_common.rs) — see the P3 plan's "Deferred" section: build the Swift
// app against this client, verify a real turn + approval roundtrip, then
// delete the FFI glue.
//
// Wire framing: one newline-terminated JSON object per message, same as
// oneai-bus's `serialize_yield` / `serialize_directive`. See
// crates/oneai-bus/src/protocol.rs for the canonical `kind` tags.
//
// This is a skeleton — connect/send/receive/approval loop only. Wire it into
// the SwiftUI app's view model; the app renders off the yields it receives.

import Foundation
import Network

/// One yield the engine sent, decoded enough to dispatch on `kind`.
public struct EngineYield {
    public let kind: String
    public let json: [String: Any]
}

public protocol OneAIBusClientDelegate: AnyObject {
    /// Called for every yield. Special-case `approval_request` — the delegate
    /// should present UI and call `client.respondToApproval(...)`; everything
    /// else renders into the chat surface.
    func busClient(_ client: OneAIBusClient, didReceive yield: EngineYield)
    /// Called when the socket closed or failed.
    func busClient(_ client: OneAIBusClient, didCloseWithError error: Error?)
}

/// Network.framework connection to `oneai serve` over a Unix domain socket.
public final class OneAIBusClient {
    private let socketPath: String
    private var connection: NWConnection?
    public weak var delegate: OneAIBusClientDelegate?

    /// - Parameter socketPath: filesystem path to the UDS (default
    ///   `~/.oneai/serve.sock`). Pass the same value as `oneai serve --socket`.
    public init(socketPath: String = "~/.oneai/serve.sock") {
        let expanded = socketPath.withCString { strlen($0) > 0 && socketPath.hasPrefix("~")
            ? NSString(string: socketPath).expandingTildeInPath
            : socketPath }
        self.socketPath = expanded
    }

    public func connect() {
        let endpoint = NWEndpoint.unix(path: socketPath)
        let parameters = NWParameters.tcp
        // NWParameters default works for UDS; .unix path is set on the endpoint.
        let conn = NWConnection(to: endpoint, using: parameters)
        self.connection = conn
        conn.stateUpdateHandler = { [weak self] state in
            switch state {
            case .ready:
                self?.startReceiveLoop()
            case .failed(let err):
                self?.delegate?.busClient(self!, didCloseWithError: err)
            case .cancelled:
                self?.delegate?.busClient(self!, didCloseWithError: nil)
            default:
                break
            }
        }
        conn.start(queue: .global(qos: .userInitiated))
    }

    public func disconnect() {
        connection?.cancel()
    }

    // MARK: - Sending directives

    /// Send a `Directive` as one JSON line. `payload` is the full object
    /// including its `kind` field, e.g. `{"kind": "user_message", "content": [{"type":"text","text":"hi"}]}`.
    public func sendDirective(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              var line = String(data: data, encoding: .utf8) else { return }
        if !line.hasSuffix("\n") { line.append("\n") }
        guard let bytes = line.data(using: .utf8) else { return }
        connection?.send(content: bytes, completion: .contentProcessed { err in
            if let err = err { print("[OneAIBusClient] send failed: \(err)") }
        })
    }

    /// Convenience: start a turn.
    public func sendUserMessage(_ text: String) {
        sendDirective([
            "kind": "user_message",
            "content": [["type": "text", "text": text]],
        ])
    }

    /// Convenience: interrupt the in-flight turn.
    public func sendInterrupt(reason: String) {
        sendDirective([
            "kind": "interrupt",
            "reason": ["Custom": ["reason": reason]],
        ])
    }

    /// Reply to an `approval_request` yield. `proceed == true` sends
    /// `InteractionResponse::Proceed` (a bare `"Proceed"` JSON string — see
    /// oneai-core's externally-tagged enum).
    public func respondToApproval(requestId: String, proceed: Bool) {
        let response: Any
        if proceed {
            // Unit variant → bare string "Proceed" (externally-tagged serde).
            response = "Proceed"
        } else {
            response = ["Abort": ["reason": "user denied"]]
        }
        sendDirective([
            "kind": "approve",
            "request_id": requestId,
            "response": response,
        ])
    }

    // MARK: - Receiving yields

    private var lineBuffer = Data()

    private func startReceiveLoop() {
        connection?.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) {
            [weak self] data, _, _, error in
            guard let self = self else { return }
            if let data = data, !data.isEmpty {
                self.lineBuffer.append(data)
                self.drainCompleteLines()
            }
            if let error = error {
                self.delegate?.busClient(self, didCloseWithError: error)
                return
            }
            // Keep reading until the connection closes.
            self.startReceiveLoop()
        }
    }

    private func drainCompleteLines() {
        while let nl = lineBuffer.firstIndex(of: 0x0A) {
            var lineData = lineBuffer.prefix(upTo: nl)
            // Strip trailing \r if present.
            if lineData.last == 0x0D { lineData = lineData.dropLast() }
            lineBuffer.removeSubrange(lineBuffer.startIndex...nl)
            guard let obj = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any],
                  let kind = obj["kind"] as? String else { continue }
            let yield = EngineYield(kind: kind, json: obj)
            DispatchQueue.main.async { [weak self] in
                guard let self = self else { return }
                self.delegate?.busClient(self, didReceive: yield)
            }
        }
    }
}

File: examples/native/windows/OneAIBusClient.cs
// OneAIBusClient.cs
// Windows native frontend for `oneai serve` (the engine-bus sidecar).
//
// A Directive writer + Yield reader over a named pipe. This replaces the
// hand-written C# extern-C facade against oneai-uniffi's c_facade.rs — see
// the P3 plan's "Deferred" section: build the WinUI3 app against this client,
// verify a real turn + approval roundtrip, then collapse c_facade.rs to the
// mobile-only 3-symbol form (P4).
//
// Wire framing: one newline-terminated JSON object per message, same as
// oneai-bus's `serialize_yield` / `serialize_directive`. See
// crates/oneai-bus/src/protocol.rs for the canonical `kind` tags.
//
// On Windows the sidecar listens on a named pipe. The pipe NAME is the
// flattened socket path (`oneai_supervisor::transport::to_pipe_name`). The
// cleanest cross-plate default: start the sidecar with
//   `oneai serve --socket oneai-serve`
// which on Windows becomes the pipe `\\.\pipe\oneai-serve`. Pass that name
// (without the `\\.\pipe\` prefix) to the client, or pass the full path — the
// client normalizes both.
//
// This is a skeleton — connect/send/receive/approval loop only.

using System;
using System.IO;
using System.IO.Pipes;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace OneAI.Native;

/// <summary>One yield the engine sent, decoded enough to dispatch on <c>kind</c>.</summary>
public sealed class OneAIBusClient : IDisposable
{
    private readonly string _pipeName;
    private NamedPipeClientStream? _pipe;
    private StreamReader? _reader;
    private StreamWriter? _writer;
    private CancellationTokenSource? _cts;

    /// <param name="pipeName">Pipe name without the <c>\\.\pipe\</c> prefix
    /// (e.g. <c>oneai-serve</c>), or the full path.</param>
    public OneAIBusClient(string pipeName = "oneai-serve")
    {
        const string prefix = @"\\.\pipe\";
        _pipeName = pipeName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)
            ? pipeName[prefix.Length..]
            : pipeName;
    }

    public event Action<JsonElement>? OnYield;
    public event Action<Exception?>? OnClosed;

    public async Task ConnectAsync(CancellationToken ct = default)
    {
        _pipe = new NamedPipeClientStream(
            ".",
            _pipeName,
            PipeDirection.InOut,
            PipeOptions.Asynchronous);
        await _pipe.ConnectAsync(ct);

        _reader = new StreamReader(_pipe, new UTF8Encoding(false));
        _writer = new StreamWriter(_pipe, new UTF8Encoding(false)) { NewLine = "\n", AutoFlush = true };
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);

        _ = Task.Run(ReceiveLoop, _cts.Token);
    }

    public void Dispose()
    {
        _cts?.Cancel();
        _writer?.Dispose();
        _reader?.Dispose();
        _pipe?.Dispose();
        GC.SuppressFinalize(this);
    }

    // ── Sending directives ────────────────────────────────────────────────

    /// <summary>Send a <c>Directive</c> as one JSON line.</summary>
    public async Task SendDirectiveAsync(object directive, CancellationToken ct = default)
    {
        if (_writer is null) throw new InvalidOperationException("not connected");
        var json = JsonSerializer.Serialize(directive);
        await _writer.WriteLineAsync(json.AsMemory(), ct);
        await _writer.FlushAsync(ct);
    }

    public Task SendUserMessageAsync(string text, CancellationToken ct = default) =>
        SendDirectiveAsync(new
        {
            kind = "user_message",
            content = new[] { new { type = "text", text } },
        }, ct);

    public Task SendInterruptAsync(string reason, CancellationToken ct = default) =>
        SendDirectiveAsync(new
        {
            kind = "interrupt",
            reason = new { Custom = new { reason } },
        }, ct);

    /// <summary>Reply to an <c>approval_request</c> yield.</summary>
    public Task RespondToApprovalAsync(string requestId, bool proceed, CancellationToken ct = default)
    {
        // InteractionResponse is externally-tagged: the Proceed unit variant
        // serializes as the bare string "Proceed"; Abort as an object.
        object response = proceed ? (object)"Proceed" : new { Abort = new { reason = "user denied" } };
        return SendDirectiveAsync(new
        {
            kind = "approve",
            request_id = requestId,
            response,
        }, ct);
    }

    // ── Receiving yields ──────────────────────────────────────────────────

    private async Task ReceiveLoop()
    {
        var token = _cts!.Token;
        try
        {
            while (!token.IsCancellationRequested)
            {
                var line = await _reader!.ReadLineAsync(token);
                if (line is null) { OnClosed?.Invoke(null); break; }
                if (string.IsNullOrWhiteSpace(line)) continue;
                try
                {
                    using var doc = JsonDocument.Parse(line);
                    OnYield?.Invoke(doc.RootElement.Clone());
                }
                catch (JsonException ex)
                {
                    OnYield?.Invoke(JsonSerializer.SerializeToElement(new
                    {
                        kind = "error",
                        recoverable = true,
                        message = $"malformed yield line: {ex.Message}",
                    }));
                }
            }
        }
        catch (OperationCanceledException) { /* shutting down */ }
        catch (Exception ex) { OnClosed?.Invoke(ex); }
    }
}
